website/content/docs/rfcs/efficient-search-components.mdx

```diff
@@ -0,0 +1,341 @@
+```
+
+As for the indexes available, they should be provided through a specific endpoint.
+
+## Rationale and Alternatives
+
+### Node Keys
+
+The current implementation uses a simple string-based key for each node, which is sufficient for the initial use case. However, this could be extended to support more complex key structures in the future. Strings keys has the limitation of not ordering integer values.
+
+### Concurrency Control
+
+The current implementation uses tree-level locking with a single `RWMutex` per B+ tree instance. This approach was chosen for its simplicity and correctness guarantees, but it can become a performance bottleneck under high concurrent workloads.
+
+**Alternative Considered: Per-Node Locking**
+
+A more sophisticated approach would implement fine-grained locking at the node level, where each node maintains its own mutex. This would allow multiple concurrent operations to proceed simultaneously as long as they operate on different parts of the tree.
+
+```go
+type Node struct {
+    // ... existing fields
+    lock sync.RWMutex
+}
+```
+
+The tree-level locking approach was selected for the initial implementation to ensure correctness and maintainability. Per-node locking could be considered as a future optimization if benchmarking identifies concurrency as a bottleneck.
+
+### Storage
+
+The current storage system has the limitation of a one-to-one mapping between keys and nodes. This means that each key can only have one node associated with it, which can be a bottleneck when dealing with large amounts of data. To overcome this limitation, a new storage system that supports multiple nodes per key should be implemented.
+
+**Alternative Considered: Page-Based Storage**
+A page-based storage system would allow multiple nodes to be associated with a single key, enabling more efficient storage and retrieval of data. This would require a more complex implementation but would provide better performance for large datasets. The current implementation uses a simple key-value store, which is sufficient for the initial use case but may need to be replaced in the future as the system scales.
+
+## Downsides
+
+Code complexity and additional load on the system and storage when building indexes. It is suggested to make index building opt-in and also tunable so that system administrators can disable index building for mounts if desired.
+
+## Security Implications
+
+No security implications were identified.
+
+## User/Developer Experience
+
+With these constructs available, an OpenBao developer will be able to dynamically configure and build indexes on the data persisted in storage.
+No user facing changes are expected with this RFC as it only introduces the constructs to build the indexes.
+
+## Unresolved Questions
+
+- How to keep track of the current progress of an index that is being incrementally built;
+- Purging or removing all nodes associated with a key from multiple trees;
```

Comment:
I think you have to just use a lock/transaction across the index to do this?

Reply:
Yes, now that I think better about it something like that would work. It's somewhat the same as querying multiple trees and joining their results.

Reply:
Can you clarify what do you mean with that?

Reply:
Say I had multiple indices over a KVv2 entry:

```json
{
 "value": "...",
 "author": "...",
 "security_level": "..."
}
```

and I call `DELETE /data/my_entry`, causing me to want to remove `my_entry` from all three indices above. I'd do:

```go
txn, err := req.Storage.BeginTransaction()
storage, err := bptree.NewStorage(ctx, txn)

for indexName, indexValue := range getIndicesInEntry("my-entry") {
  idx := b.idxMgr.GetIndex(ctx, storage, indexName)
  idx.Delete(ctx, storage, indexValue)
  // or
  idx.Delete(ctx, storage, "my-entry")
}

txn.Commit()
```

Reply:
Yes, the current implementation would work something like that but has its limitations. If operating with an "external" transaction, we don't know whether or not an operation was successful until is effectively committed. For that reason, I have decided to disable caching if operating on an logical.storage transaction. I'm working on some storage improvements, that will be pushing the PoC branch soon. Worth updating the spec?

Reply:
@Gabrielopesantos Right, I had really wished the physical cache system allowed snapshotting, but alas. :(  What we ended up doing was creating a new (smaller) cache for the transaction, so at least inside the transaction you can take advantage of the benefits. 

Which cache library are you using? If you can get a soft-clone, that'd be ideal. Even if you discard the state afterwards, that'd be fine.

Reply:
> Which cache library are you using?

The one created in  `sdk/lru/lru.go`, its just a wrapper around `github.com/hashicorp/golang-lru/v2`.

> If you can get a soft-clone, that'd be ideal. Even if you discard the state afterwards, that'd be fine.

Yes, something like that should be feasible. Going to look into that.